transaction/core/src/ring_signature/rct_bulletproofs.rs

```diff
@@ -269,19 +270,26 @@ impl SignatureRctBulletproofs {
             for (token_id, range_proof) in token_ids.iter().zip(self.range_proofs.iter()) {
                 let generator = generator_cache.get(*token_id);
 
-                let commitments: Vec<CompressedRistretto> = self
-                    .pseudo_output_commitments
-                    .iter()
-                    .zip(self.pseudo_output_token_ids.iter())
-                    .chain(output_commitments.iter().zip(self.output_token_ids.iter()))
-                    .filter_map(|(compressed_commitment, this_token_id)| {
-                        if token_id == this_token_id {
-                            Some(compressed_commitment.point)
-                        } else {
-                            None
-                        }
-                    })
-                    .collect();
+                let commitments: Vec<CompressedRistretto> = zip_exact(
+                    self.pseudo_output_commitments.iter(),
+                    self.pseudo_output_token_ids.iter(),
+                )
+                .chain(zip_exact(
+                    output_commitments.iter(),
+                    self.output_token_ids.iter(),
+                ))
+                .filter_map(|result| {
+                    result
+                        .map(|(compressed_commitment, this_token_id)| {
+                            if token_id == this_token_id {
+                                Some(compressed_commitment.point)
+                            } else {
+                                None
+                            }
+                        })
+                        .transpose()
+                })
+                .collect::<Result<_, ZipExactError>>()?;
```

Comment:
How do people feel about this code? Is this readable?

I feel like I'm not very happy with it, the need to deal with errors returned by the iterator seems to cause a syntax explosion.

Maybe it's better to just use a simple for loop than to try to use iter and collect here?

```
for result in zip_exact(...).chain(zip_exact(...)) {
    let (compressed_commtiment, this_token_id) = result?;

    if token_id == this_token_id {
       compressed_commitments.push(...)
    }
}
```

LMK what you think

Reply:
It feels like in most langs including rust, there's a drive to say "shorter, more language idiomatic code is better" but sometimes it's not in these type of cases where other factors come into play, and I feel like here a clear for loop IS better.